### PR TITLE
Implemented screenshot API

### LINKED ISSUE #31: Add API to take a screenshot

### PR DESCRIPTION
**Pull request by eugene-korobko**
#Fixes 31

**enhancement** 

**PR checklist:**

- [x] Addresses an existing issue: fixes #31
- [x] Includes tests: graphic test 
- [x] Documentation update: js docs for new method

**Overview of change:**
Added a method takeScreenshot on IChartApi to get image of the whole chart as a single image

### CODE REVIEW

File: src/gui/chart-widget.ts [hunk 1]
```diff
@@ -204,6 +205,82 @@ export class ChartWidget implements IDestroyable {
 		this._paneWidgets[0].disableBranding();
 	}
 
+	public takeScreenshot(): string {
+		// calculate target size
```
Comment (timocov) [inline src/gui/chart-widget.ts]: If `_invalidateMask` is not `null` we need to call `_drawImpl` here to validate the state before generating screenshot.

File: src/gui/pane-widget.ts [hunk 2]
```diff
@@ -26,6 +26,12 @@ export interface HitTestResult {
 	view: IPaneView;
 }
 
+const enum LogoConstants {
+	FONT_SIZE = '7pt',
+	LEFT_MARGIN = 6,
+	BOTTOM_MARGIN = 4,
```
Comment (timocov) [inline src/gui/pane-widget.ts]: I believe we can use CamelCase naming here (e.g. `BottomMargin`, `LeftMargin` and `FontSize`)

File: src/gui/pane-widget.ts [hunk 3]
```diff
@@ -26,6 +26,12 @@ export interface HitTestResult {
 	view: IPaneView;
 }
 
+const enum LogoConstants {
+	FONT_SIZE = '7pt',
```
Comment (timocov) [inline src/gui/pane-widget.ts]: We can't use `pt` for rendering on canvas. We should use `px` for canvas and `pt` for DOM element to avoid hidpi rendering problems.

File: src/gui/pane-widget.ts [hunk 4]
```diff
@@ -435,6 +441,24 @@ export class PaneWidget implements IDestroyable {
...
+		resizeCanvas(res, this._size);
+		const ctx = ensureNotNull(getContext2d(res));
+		ctx.drawImage(this._canvas, 0, 0);
+		if (this._brandingElement !== null) {
+			const dpr = window.devicePixelRatio || 1;
+			// one point is 1/72 of inch
+			// standard resolution is 96 dime per inch
+			const fontSize = (LogoConstants.FONT_SIZE / 72) * 96 * dpr;
+			ctx.fillStyle = colorWithTransparency(this._chart.options().layout.textColor, 0.9);
+			ctx.font = `${fontSize}px ${defaultFontFamily}`;
```
Comment (timocov) [inline src/gui/pane-widget.ts]: Please use `makeFont`

File: src/api/chart-api.ts [hunk 5]
```diff
@@ -255,6 +255,10 @@ export class ChartApi implements IChartApi, DataUpdatesConsumer<SeriesType> {
 		this._chartWidget.disableBranding();
 	}
 
+	public takeScreenshot(): string {
+		return this._chartWidget.takeScreenshot();
```
[the remark this answers is not in the document]
Comment (timocov) [reply src/api/chart-api.ts]: Agreed.

File: src/api/ichart-api.ts [hunk 6]
```diff
@@ -147,4 +147,10 @@ export interface IChartApi {
 	 * Please read the description of this method in the documentation to learn more about the conditions of branding removal.
 	 */
 	disableBranding(): void;
+
+	/**
+	 * Make a screenshot of the chart with all the elements excluding crosshair.
+	 * @Return - a result in format data url
```
[the remark this answers is not in the document]
Comment (eugene-korobko) [reply src/api/ichart-api.ts]: Fixed

File: src/gui/chart-widget.ts [hunk 7]
```diff
@@ -204,6 +205,82 @@ export class ChartWidget implements IDestroyable {
...
+		const firstPane = this._paneWidgets[0];
+		let targetWidth = firstPane.getSize().w;
+		if (this._options.priceScale.position !== 'none') {
+			targetWidth += ensureNotNull(firstPane.priceAxisWidget()).getWidth();
+		}
+		let targetHeight = 0;
+		for (const paneWidget of this._paneWidgets) {
+			targetHeight += paneWidget.getSize().h;
```
[the remark this answers is not in the document]
Comment (eugene-korobko) [reply src/gui/chart-widget.ts]: fixed

File: src/gui/pane-widget.ts [hunk 8]
```diff
@@ -435,6 +441,24 @@ export class PaneWidget implements IDestroyable {
...
+			this._drawGrid(ctx);
+			this._drawWatermark(ctx);
+			this._drawSources(ctx);
+		}
+	}
+
+	public drawBranding(ctx: CanvasRenderingContext2D): void {
+		if (this._brandingElement !== null) {
+			ctx.fillStyle = colorWithTransparency(this._chart.options().layout.textColor, 0.9);
+			ctx.font = `${LogoConstants.FONT_SIZE} ${defaultFontFamily}`;
```
[the remark this answers is not in the document]
Comment (eugene-korobko) [reply src/gui/pane-widget.ts]: fixed

File: src/gui/chart-widget.ts [hunk 9]
```diff
@@ -204,6 +205,82 @@ export class ChartWidget implements IDestroyable {
...
+			this._timeAxisWidget.drawOnCanvas(ctx);
+			ctx.restore();
+			if (this._options.priceScale.position === 'right') {
+				targetX = firstPane.getSize().w;
+				drawStub();
+				ctx.restore();
+			}
+		}
+
+		// draw branding
```
[the remark this answers is not in the document]
Comment (eugene-korobko) [reply src/gui/chart-widget.ts]: no, it is just a comment. Removed now. Panes do drawing branding

File: src/gui/chart-widget.ts (hunk 1 above)
Comment (eugene-korobko) [reply src/gui/chart-widget.ts]: done

File: src/gui/pane-widget.ts (hunk 2 above)
Comment (eugene-korobko) [reply src/gui/pane-widget.ts]: done

File: src/api/chart-api.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (eugene-korobko) [reply src/api/chart-api.ts]: changed

File: src/gui/price-axis-widget.ts [hunk 10]
```diff
@@ -265,23 +265,28 @@ export class PriceAxisWidget implements IDestroyable {
 		model.resetPriceScale(pane, ensureNotNull(this.priceScale()));
 	}
 
+	public drawOnCanvas(ctx: CanvasRenderingContext2D): void {
```
Comment (timocov) [inline src/gui/price-axis-widget.ts]: Do we still need this method?

File: tests/e2e/graphics/test-cases/take-screenshot.js [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+		drawBorder: true,
+		borderColor: 'blue',
+	});
+
+	mainSeries.setData(generateData());
+
+	// create canvas to draw screenshot
+	chart.resize(240, 600, true);
+
+	var screenshot = chart.takeScreenshot().toDataURL();
```
Comment (timocov) [inline tests/e2e/graphics/test-cases/take-screenshot.js]: If we return canvas element, maybe we can just insert it as DOM element without creating `img` element?

File: src/gui/time-axis-widget.ts [hunk 12]
```diff
@@ -232,18 +240,18 @@ export class TimeAxisWidget implements MouseEventHandlers, IDestroyable {
...
-		this._drawCrosshairLabel(this._topCanvasContext);
+		this._drawTickMarks(this._canvasContext);
+		this._drawBackLabels(this._canvasContext);
 
 		if (this._stub !== null) {
 			this._stub.paint(type);
 		}
+
+		this._chart.model().crosshairSource().updateAllViews();
+		this._drawCrosshairLabel(this._topCanvasContext);
```
Comment (timocov) [inline src/gui/time-axis-widget.ts]: It looks like we can revert this changes in `paint` method at all?

File: src/gui/chart-widget.ts [hunk 13]
```diff
@@ -204,6 +205,86 @@ export class ChartWidget implements IDestroyable {
...
+		resizeCanvas(targetCanvas, new Size(targetWidth, targetHeight));
+		const ctx = ensureNotNull(getContext2d(targetCanvas));
+		let targetX = 0;
+		let targetY = 0;
+
+		const drawPriceAxises = () => {
+			for (const paneWidget of this._paneWidgets) {
+				const image = ensureNotNull(paneWidget.priceAxisWidget()).getImage();
+				ctx.drawImage(image, targetX, targetY);
+				targetY += paneWidget.getSize().h;
```
Comment (timocov) [inline src/gui/chart-widget.ts]: It looks like there should be pane separator's height as well, not only pane's height?

File: src/gui/chart-widget.ts [hunk 14]
```diff
@@ -204,6 +205,86 @@ export class ChartWidget implements IDestroyable {
 		this._paneWidgets[0].disableBranding();
 	}
 
+	public takeScreenshot(): HTMLCanvasElement {
+		if (this._invalidateMask !== null) {
+			this._drawImpl(this._invalidateMask);
+			this._invalidateMask = null;
+		}
+		// calculate target size
```
Comment (timocov) [inline src/gui/chart-widget.ts]: Can we use `this._options.width` and `this._options.height` here?

File: src/gui/pane-separator.ts [hunk 15]
```diff
@@ -66,10 +69,25 @@ export class PaneSeparator implements IDestroyable {
 		}
 	}
 
+	public getHeight(): number {
+		return SEPARATOR_HEIGHT;
```
Comment (timocov) [inline src/gui/pane-separator.ts]: It looks like this constant is already exported and used in `ChartWidget::_adjustSizeImpl`. Maybe we can just used it in `takeScreenshot` method as well without adding `PaneSeparator::getHeight` method?

File: src/gui/pane-widget.ts [hunk 16]
```diff
@@ -435,6 +441,24 @@ export class PaneWidget implements IDestroyable {
...
+		resizeCanvas(res, this._size);
+		const ctx = ensureNotNull(getContext2d(res));
+		ctx.drawImage(this._canvas, 0, 0);
+		if (this._brandingElement !== null) {
+			const dpr = window.devicePixelRatio || 1;
+			// one point is 1/72 of inch
+			// standard resolution is 96 dime per inch
+			const fontSize = (LogoConstants.FontSize / 72) * 96 * dpr;
+			ctx.fillStyle = colorWithTransparency(this._chart.options().layout.textColor, 0.9);
+			ctx.font = makeFont(fontSize, defaultFontFamily);
```
Comment (timocov) [inline src/gui/pane-widget.ts]: The fallback for the second arg of `makeFont` is `defaultFontFamily` already. Maybe we can just strip it here?

File: src/gui/price-axis-stub.ts [hunk 17]
```diff
@@ -104,33 +108,33 @@ export class PriceAxisStub implements IDestroyable {
 		return this._isLeft;
 	}
 
-	private _drawBorder(): void {
+	private _drawBorder(ctx: CanvasRenderingContext2D): void {
```
Comment (timocov) [inline src/gui/price-axis-stub.ts]: It looks like we can revert changes in this file except adding the new method `getImage`

File: src/gui/pane-widget.ts [hunk 18]
```diff
@@ -435,6 +441,24 @@ export class PaneWidget implements IDestroyable {
...
+	public getImage(): HTMLCanvasElement {
+		const res = document.createElement('canvas');
+		resizeCanvas(res, this._size);
+		const ctx = ensureNotNull(getContext2d(res));
+		ctx.drawImage(this._canvas, 0, 0);
+		if (this._brandingElement !== null) {
+			const dpr = window.devicePixelRatio || 1;
+			// one point is 1/72 of inch
+			// standard resolution is 96 dime per inch
+			const fontSize = (LogoConstants.FontSize / 72) * 96 * dpr;
```
Comment (timocov) [inline src/gui/pane-widget.ts]: Hm, but hidpi canvas already multiply font size in `devicePixelRatio`, why we need to do it outside one more time? Can we just use `LogoConstants.FontSize`px?

File: src/api/ichart-api.ts [hunk 19]
```diff
@@ -147,4 +147,10 @@ export interface IChartApi {
 	 * Please read the description of this method in the documentation to learn more about the conditions of branding removal.
 	 */
 	disableBranding(): void;
+
+	/**
+	 * Make a screenshot of the chart with all the elements excluding crosshair.
+	 * @return a canvas with the chart drawn on
```
Comment (timocov) [inline src/api/ichart-api.ts]: `@returns` (with the `s` letter in the end)

File: src/gui/price-axis-widget.ts (hunk 10 above)
Comment (eugene-korobko) [reply src/gui/price-axis-widget.ts]: removed

File: tests/e2e/graphics/test-cases/take-screenshot.js (hunk 11 above)
Comment (eugene-korobko) [reply tests/e2e/graphics/test-cases/take-screenshot.js]: ok

File: src/gui/time-axis-widget.ts (hunk 12 above)
Comment (eugene-korobko) [reply src/gui/time-axis-widget.ts]: removed

File: src/gui/chart-widget.ts (hunk 14 above)
Comment (eugene-korobko) [reply src/gui/chart-widget.ts]: ok

File: src/gui/pane-separator.ts (hunk 15 above)
Comment (eugene-korobko) [reply src/gui/pane-separator.ts]: removed

File: src/gui/pane-widget.ts (hunk 16 above)
Comment (eugene-korobko) [reply src/gui/pane-widget.ts]: ok

File: src/gui/price-axis-stub.ts (hunk 17 above)
Comment (eugene-korobko) [reply src/gui/price-axis-stub.ts]: ok

File: src/gui/pane-widget.ts (hunk 18 above)
Comment (eugene-korobko) [reply src/gui/pane-widget.ts]: we have to keep converting from points to pixels, removed devicePixelRatio

File: src/api/ichart-api.ts (hunk 19 above)
Comment (eugene-korobko) [reply src/api/ichart-api.ts]: ok

File: src/gui/time-axis-widget.ts [hunk 20]
```diff
@@ -232,14 +240,11 @@ export class TimeAxisWidget implements MouseEventHandlers, IDestroyable {
...
-		const ctx = this._canvasContext;
-		this._drawBackground(ctx);
-		this._drawBorder(ctx);
+		this._drawBackground(this._canvasContext);
+		this._drawBorder(this._canvasContext);
 
-		this._drawTickMarks(ctx);
-		this._drawBackLabels(ctx);
-		this._chart.model().crosshairSource().updateAllViews();
-		this._drawCrosshairLabel(this._topCanvasContext);
```
Comment (timocov) [inline src/gui/time-axis-widget.ts]: Why this code is removed? Can we just revert this method to the master's one?

File: src/gui/pane-widget.ts (hunk 18 above)
Comment (timocov) [reply src/gui/pane-widget.ts]: I hope this code will be removed in the branch with removing the branding :)

File: src/gui/chart-widget.ts (hunk 13 above)
Comment (eugene-korobko) [reply src/gui/chart-widget.ts]: this code has been rewritten

Comment (timocov) [reply src/gui/chart-widget.ts]: Where? It looks like there is no adding `SEPARATOR_HEIGHT` to the `targetY` variable, or I miss something?

**Comment by Business223rx** [comment]
I can't sign in to github

**Comment by hazhir1811** [comment]
Hello how can we use this method on the chart?

**Comment by timocov** [comment]
See https://github.com/tradingview/lightweight-charts/blob/dd7fb362a646bc8021b050b32f2dca233baf5e12/src/api/ichart-api.ts#L150-L154 for API reference or https://github.com/tradingview/lightweight-charts/blob/ad8326c28b2d38c0de445eab172a6a63688ea7ff/tests/e2e/graphics/test-cases/take-screenshot.js#L43-L48 for usage in our test cases.

We need to add docs for this method...
